### DISCUSSION TITLE
AlpineJS and Blazor 8 Enhanced navigation

### DISCUSSION THREAD

**Discussion by oliverw** [alpinejs/alpine#3887] [1. Help]
When enabling [Enhanced navigation](https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-8.0#enhanced-navigation-and-form-handling) in Blazor clicking an anchor results in the framework intercepting the request and performing a fetch request instead. Blazor then patches the response content into the page's DOM. I'd say this would cause problems since `Alpine.start()` only runs on the first full page load. It is possible to react when Blazor has patched the DOM:

```
Blazor.addEventListener('enhancedload', () => {
  // how to tell Alpine to update from changed dom?
});
```

I've read this [thread regarding liveview](https://github.com/alpinejs/alpine/issues/282) and there's a lot of talk about using `Alpine.discoverUninitializedComponents` but I really didn't find any conclusive answer.

**Comment by SimoTod** [comment]
Have you tried? That post was about alpine v2. Alpine v3 is different, if the framework patches the Dom by adding/removing elements on a page where Alpine has already been started, Alpine should see the changes and initialise the new elements.

**Reply by oliverw** [reply]
No I haven't tried it yet. I've posted this to figure out if there's a way that both frameworks play nice. 

> Alpine should see the changes and initialise the new elements.

How is that even possible if I may ask? Mutation observers?

**Reply by oliverw** [reply]
Seems to work so far 👍

**Reply by ekwoka** [reply]
Yes, there is a mutation observer. You can just add and remove attributes or add and remove elements and Alpine sees it and initializes or cleansup whatever has changed.

**Reply by cdock1029** [reply]
@oliverw did you ever get this to work? With SSR and enhanced nav, the state of the dom elements gets reset. Debugging in the console, the Alpine state is still correct. If I have a x-show, and that value is false, it get's revealed when I navigate or submit a form.

**Reply by cdock1029** [reply]
So if I'm observing correctly here, Blazor enhanced nav does a fetch request for the next page, and returns the entire new page as full html in the fetch response (then diffs this with the library client side). Is this something Alpine is not designed for (as opposed to more surgical updates of html partials)? I'm trying to keep a popup hidden in my layout that wraps my page content, I assumed it would maintain this between navigations. Then again HTMX does this full page fetch by default too I think, so not sure how Alpine is handled there.

**Reply by oliverw** [reply]
@cdock1029 The answer is "kind of". But it's a neverending struggle. Avoid it if you can.

**Reply by cdock1029** [reply]
Reading here, this "blowing away of the dom completely" is a broad pain point 

https://github.com/dotnet/aspnetcore/issues/42561#issuecomment-2722255871

**Reply by cdock1029** [reply]
> @cdock1029 The answer is "kind of". But it's a neverending struggle. Avoid it if you can.

Avoid which one? 😂

**Reply by oliverw** [reply]
@cdock1029 Blazor + AlpineJS, or anything that mutates the DOM.

**Reply by cdock1029** [reply]
@oliverw 

For future reference in case you or others don't know, this solved my Blazor SSR issue entirely (preserve certain dom nodes so Alpine can maintain its visual state for them):

`data-permanent`

Found here: 

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-9.0

> Blazor's enhanced navigation and form handing may undo dynamic changes to the DOM if the updated content isn't part of the server rendering. To preserve the content of an element, use the data-permanent attribute.

**Comment by reecelikesramen** [comment]
Working on this in #4714 and made some progress you might be interested in

**Comment by davidwebca** [comment]
Following this thread as I'm battling this issue right now.
